workflow/controller/dag.go

```diff
@@ -695,7 +695,20 @@ func (d *dagContext) evaluateDependsLogic(taskName string) (bool, bool, error) {
 		}
 
 		// If a task happens to have an onExit node, don't proceed until the onExit node is fulfilled
-		if onExitNode := d.wf.GetNodeByName(common.GenerateOnExitNodeName(taskName)); onExitNode != nil {
+		if onExitNode := d.wf.GetNodeByName(common.GenerateOnExitNodeName(depNode.Name)); onExitNode != nil {
+			if !onExitNode.Fulfilled() {
+				return false, false, nil
+			}
+		}
+
+		// Previously we used `depNode.DisplayName` to generate all onExit node names. However, as these can be non-unique
+		// we transitioned to using `depNode.Name` instead, which are guaranteed to be unique. In order to not disrupt
+		// running workflows during upgrade time, we do an additional check to see if there is an onExit node with the old
+		// name (`depNode.DisplayName`).
+		// TODO: This scaffold code should be removed after a couple of "grace period" version upgrades to allow transitions. It was introduced in v3.0.0
+		// See more: https://github.com/argoproj/argo-workflows/issues/5502
+		legacyOnExitNodeName := common.GenerateOnExitNodeName(depNode.DisplayName)
+		if onExitNode := d.wf.GetNodeByName(legacyOnExitNodeName); onExitNode != nil && d.wf.GetNodeByName(depNode.Name).HasChild(onExitNode.ID) {
```

Comment:
This is tested by `TestOnExitDAGStatusCompatibility` in this PR

Reply:
nice